MsgPack bin8/bin16/bin32 support

Sanae6 opened the pull request:
I've added binary support to the msgpack deserializer and serializer, along with a fallback for the json serializer. I've also added some tests to the relevant files. Let me know if there's anything I can do to make this implementation better. Thank you for the amazing library :)

This would fix and close #922.

Sanae6 committed:
MsgPack bin8/bin16/bin32 draft

Sanae6 committed:
Remove tests that were broken by bin impl

bblanchon commented:
Hi @Sanae6,

Thank you for this excellent PR!

Before starting the review, I'd like you to remove `LinkedBinary`.
Indeed, binary values should be treated like raw JSON strings and always be copied; this is part of a larger (incomplete) refactoring to reduce the size of `VariantContent`.
Then, of course, simplify the names accordingly, since we no longer need the distinction between "owned" and "linked" binary values.

Best regards,
Benoit

Sanae6 committed:
Replace OwnedBinaryValue/LinkedBinaryValue with BinaryValue

Signed-off-by: Aubrey (Sanae) <aubrey@hall.ly>

Sanae6 commented:
I removed LinkedBinaryValue and renamed OwnedBinaryValue to BinaryValue. Not sure what to name [Binary](https://github.com/bblanchon/ArduinoJson/blob/110e63981c4871217d42981b31aaa76abf1b597e/src/ArduinoJson/MsgPack/BinaryValue.hpp#L79), but I left it as something sensible enough.

bblanchon commented on src/ArduinoJson/MsgPack/BinaryValue.hpp:
```diff
@@ -0,0 +1,91 @@
+using Binary = BinaryValue<detail::StringNode*>;
+
+template <typename T>
+inline BinaryValue<T> binary(const T& value) {
```
The word `binary` is too generic.
Please rename the class to `MsgPackBinary` and remove the function.

bblanchon commented on src/ArduinoJson/MsgPack/BinaryValue.hpp:
```diff
@@ -0,0 +1,91 @@
+ARDUINOJSON_BEGIN_PUBLIC_NAMESPACE
+
+template <typename T>
+class BinaryValue {
```
Supporting container is cool, but not worth the complexity.
Please simplify the class to support only `const void*` / `size_t`.

bblanchon commented on src/ArduinoJson/MsgPack/BinaryValue.hpp:
```diff
@@ -0,0 +1,91 @@
+};
+
+template <>
+class BinaryValue<detail::StringNode*> {
```
I'm surprised to see this.
Please remove this specialization.

bblanchon commented on extras/tests/MsgPackDeserializer/deserializeVariant.cpp:
```diff
@@ -139,6 +140,55 @@ TEST_CASE("deserialize MsgPack value") {
+    REQUIRE(reinterpret_cast<const char*>(binary.data())[0] == 5);
+  }
+
+#if ARDUINOJSON_STRING_LENGTH_SIZE >= 4
```
Please move this test to a dedicated file in `MixedConfiguration`, so you can control the value of `ARDUINOJSON_STRING_LENGTH_SIZE`

bblanchon commented on extras/tests/MsgPackDeserializer/errors.cpp:
```diff
@@ -134,18 +134,6 @@ TEST_CASE("deserializeMsgPack() returns IncompleteInput") {
-  SECTION("bin 32") {
-    testIncompleteInput("\xc6\x00\x00\x00\x01X", 6);
-  }
-
```
These tests should still pass.
Please restore them.

bblanchon commented on extras/tests/MsgPackDeserializer/filter.cpp:
```diff
@@ -375,52 +375,6 @@ TEST_CASE("deserializeMsgPack() filter") {
-                  Reallocate(sizeofPool(), sizeofObject(1)),
-              });
-      }
-
```
These tests should still pass.
Please restore them.

bblanchon commented on src/ArduinoJson/Configuration.hpp:
```diff
@@ -134,7 +134,7 @@
 #    define ARDUINOJSON_STRING_LENGTH_SIZE 1  // up to 255 characters
 #  else
-#    define ARDUINOJSON_STRING_LENGTH_SIZE 2  // up to 65535 characters
+#    define ARDUINOJSON_STRING_LENGTH_SIZE 4  // up to 65535 characters
```
Please revert this change.
You can change settings in the compilation units of the `MixedConfiguration` executable.

bblanchon commented on src/ArduinoJson/Json/JsonSerializer.hpp:
This feature is for message pack only; other users should not be affected.
Besides, JSON doesn't support hexadecimal.
Please revert all changes affecting the JSON related code.

bblanchon replied:
Since you have to implement `visitBinary()`, please emit `null`.
I'll tell you later how I plan to eliminate this.

bblanchon commented on src/ArduinoJson/Variant/VariantContent.hpp:
```diff
@@ -31,7 +31,8 @@ enum {
   VALUE_IS_FLOAT = 0x0C,
 
-  COLLECTION_MASK = 0x60,
+  COLLECTION_MASK = 0x70,
```
Collections in this context are arrays and objects, but not binary values.
Please restore the original value.

Sanae6 replied:
Should I move `VALUE_IS_BINARY` out of the collections section? If so, where?

bblanchon replied:
You should use the value 0x02 which used to be reserved for linked raw strings.

Sanae6 replied:
That doesn't seem right, that would make it so that binaries aren't owned by the document, which isn't true. I have it on 0x07 locally to make sure the owned flag is set right now. Is 0x02 really fine?

bblanchon replied:
You're right. It doesn't work. Sorry.

I think I oversaw this problem because my plan was to remove `VALUE_IS_BINARY` and handle binary strings with `VALUE_IS_RAW_STRING` and let the converters do all the work.
This way, we add support for MessagePack binary strings without affecting JSON users and also pave the way to support MessagePack ext values in the future.

Nevertheless, I think it's still a good stepping stone to have the `VALUE_IS_BINARY`, even if it'll soon be removed.
Please ignore my previous comment about the value, and position the enum value where it makes sense to you. Don't overthink it, since we may remove it in a future commit.

bblanchon commented on src/ArduinoJson/MsgPack/BinaryValue.hpp:
```diff
@@ -0,0 +1,91 @@
+    return data_ + size_;
+  }
+
+  size_t size_bytes() const {
```
Please rename `size_bytes()` to `size` to be consistent with the rest of the library.

Sanae6 committed:
Make requested changes and add (broken) test

Signed-off-by: Aubrey (Sanae) <aubrey@hall.ly>

Sanae6 commented on extras/tests/MixedConfiguration/use_string_length_size_4.cpp:
This test currently fails. It seems that even when I apply the string length size, it doesn't actually apply to `StringNode::length_type`. Printing `StringNode::maxLength` in the `StringNode::create` function returned `0xffff` instead of `0xffffffff`. Any advice on how to solve that?

bblanchon replied:
This is surely due to the fact that the namespace doesn't include the value of `ARDUINOJSON_STRING_LENGTH_SIZE`.
**Let me fix that.**

Each file in this folder is named according to the setting name, so this file should be named `string_length_size_4.cpp`.

Anyway, thank you again for this excellent work 👍

Sanae6 replied:
Alright! I'll leave it to you :)

bblanchon commented on extras/tests/JsonSerializer/JsonVariant.cpp:
```diff
@@ -3,6 +3,7 @@
 // MIT License
 
 #include <ArduinoJson.h>
+#include <array>
```
This isn't needed anymore.

bblanchon commented on src/ArduinoJson/Json/JsonSerializer.hpp:
```diff
@@ -81,6 +81,11 @@ class JsonSerializer : public VariantDataVisitor<size_t> {
   }
 
+  size_t visit(MsgPackBinary) {
+    formatter_.writeRaw("null");
```
I suspect that calling `visit(nullptr)` would lead to a smaller code.

bblanchon commented on src/ArduinoJson/Memory/StringNode.hpp:
```diff
@@ -11,6 +11,7 @@
 #include <ArduinoJson/Polyfills/limits.hpp>
 
 #include <stddef.h>  // offsetof
+#include <stdio.h>
```
This was probably added by accident.
Please remove it.

Sanae6 committed:
More requested clean up

Signed-off-by: Aubrey (Sanae) <aubrey@hall.ly>

bblanchon reviewed (changes requested):
I just pushed the changes to fix this issue with `string_length_size_4.cpp`.
Please rebase your branch onto `7.x`.
Note that I created `string_length_size_{1,2,4}.cpp` to check that the fix actually works, so you'll have a conflict on `string_length_size_4.cpp`.

Sanae6 commented:
Alright! Force pushed, seems nothing is broken :+1:

bblanchon commented on extras/tests/MixedConfiguration/CMakeLists.txt:
```diff
@@ -22,6 +22,7 @@ add_executable(MixedConfigurationTests
 	use_double_1.cpp
 	use_long_long_0.cpp
 	use_long_long_1.cpp
+	string_length_size_4.cpp
```
You can remove this line as `string_length_size_4.cpp` is already included.

bblanchon commented:
It seems that you lost your version of `string_length_size_4.cpp` while merging the conflict.

bblanchon reviewed (changes requested):
This looks really good.
I'll do an in-depth review tomorrow.
Thank you for this outstanding PR.

bblanchon commented on extras/tests/MsgPackSerializer/serializeVariant.cpp:
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
   }
 
+  SECTION("bin 8") {
+    static const auto bin8 = std::array<char, 4>({1, 2, 3, 4});
```
`static const`?

bblanchon commented on extras/tests/MsgPackSerializer/serializeVariant.cpp:
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
+  }
+
+  SECTION("bin 16") {
+    static const auto bin16 = std::array<char, 0x100>({1, 2, 3, 4});
```
`static const`?

bblanchon commented on extras/tests/MsgPackDeserializer/deserializeVariant.cpp:
Remember to restore the `bin 32` test that you originally wrote in `string_length_size_4.cpp` but lost after rebasing.

bblanchon commented on extras/tests/MsgPackSerializer/serializeVariant.cpp:
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
+
+  SECTION("bin 16") {
+    static const auto bin16 = std::array<char, 0x100>({1, 2, 3, 4});
+    std::string bin16string(bin16.data(), bin16.size());
```
You could simply use `std::string(0x100, '?')`; this constructor creates a string that repeats the provided character N times.
See https://en.cppreference.com/w/cpp/string/basic_string/basic_string (2)

bblanchon commented on src/ArduinoJson/MsgPack/MsgPackBinary.hpp:
```diff
@@ -0,0 +1,25 @@
+class MsgPackBinary {
+ public:
+  explicit MsgPackBinary(const void* c, size_t size) : data_(c), size_(size) {}
+  operator const void*() const {
```
I don't think the implicit conversion to `const void*` is a good idea.
Please remove it.

bblanchon commented on src/ArduinoJson/Variant/VariantCompare.hpp:
```diff
@@ -164,6 +164,12 @@ struct VariantComparer : ComparerBase {
     return reverseResult(comparer);
   }
 
+  CompareResult visit(MsgPackBinary value) {
```
I doubt this is covered by any test.

Sanae6 committed:
Update tests and make requested changes

Signed-off-by: Aubrey (Sanae) <aubrey@hall.ly>

Sanae6 commented:
Made the changes you wanted, testing MsgPackBinary comparison actually revealed my initial implementation didn't work, so I fixed that as well.

Sanae6 committed:
Format string_length_size_4.cpp

Signed-off-by: Aubrey (Sanae) <aubrey@hall.ly>

bblanchon commented on src/ArduinoJson/MsgPack/MsgPackSerializer.hpp:
```diff
@@ -123,6 +123,21 @@ class MsgPackSerializer : public VariantDataVisitor<size_t> {
+      writeByte(0xC5);
+      writeInteger(uint16_t(value.size()));
+    } else {
+      writeByte(0xC6);
```
This line is not covered by the tests.
Please add a test in `string_length_size_4.cpp`

bblanchon commented on src/ArduinoJson/Variant/VariantData.hpp:
```diff
@@ -185,6 +190,16 @@ class VariantData {
+        return MsgPackBinary(content_.asOwnedString->data,
+                             content_.asOwnedString->length);
+      default:
+        return MsgPackBinary(nullptr, 0);
```
This line is not covered by the tests.

bblanchon commented on src/ArduinoJson/Variant/VariantData.hpp:
```diff
@@ -394,6 +413,28 @@ class VariantData {
+  static void setBinary(VariantData* var, MsgPackBinary value,
+                        ResourceManager* resources) {
+    if (!var)
+      return;
```
This line is not covered by the tests.

bblanchon commented:
I added the three missing tests, squashed, and merged.

Thank you again for this excellent contribution ❤️

bblanchon commented:
This feature is available in [ArduinoJson 7.1.0](https://github.com/bblanchon/ArduinoJson/releases/tag/v7.1.0)
